src/main/java/org/apache/flink/benchmark/MultipleInputBenchmark.java

```diff
@@ -0,0 +1,161 @@
+		env.setParallelism(1);
+		env.setRestartStrategy(RestartStrategies.noRestart());
+
+		// Setting buffer timeout to 1 is an attempt to improve twoInputMapSink benchmark stability.
+		// Without 1ms buffer timeout, some JVM forks are much slower then others, making results
+		// unstable and unreliable.
+		env.setBufferTimeout(1);
+
+		long numRecordsPerInput = RECORDS_PER_INVOCATION / 2;
+		DataStreamSource<Long> source1 = env.addSource(new LongSource(numRecordsPerInput));
+		DataStreamSource<Long> source2 = env.addSource(new LongSource(numRecordsPerInput));
+		connectAndDiscard(env, source1, source2);
+
+		env.execute();
+	}
+
+	@Benchmark
+	@OperationsPerInvocation(ONE_IDLE_RECORDS_PER_INVOCATION)
+	public void multiInputOneIdleMapSink(FlinkEnvironmentContext context) throws Exception {
+
+		StreamExecutionEnvironment env = context.env;
+		env.enableCheckpointing(CHECKPOINT_INTERVAL_MS);
+		env.setParallelism(1);
+
+		QueuingLongSource.reset();
+		DataStreamSource<Long> source1 = env.addSource(new QueuingLongSource(1, ONE_IDLE_RECORDS_PER_INVOCATION - 1));
+		DataStreamSource<Long> source2 = env.addSource(new QueuingLongSource(2, 1));
+		connectAndDiscard(env, source1, source2);
+
+		env.execute();
+	}
+
+	private void connectAndDiscard(
+			StreamExecutionEnvironment env,
+			DataStreamSource<Long> source1,
+			DataStreamSource<Long> source2) {
+		MultipleInputTransformation<Long> transform = new MultipleInputTransformation<>(
+				"custom operator",
+				new MultiplyByTwoOperatorFactory(),
+				BasicTypeInfo.LONG_TYPE_INFO,
+				1);
+
+		transform.addInput(source1.getTransformation());
+		transform.addInput(source2.getTransformation());
+
+		env.addOperator(transform);
+		new MultipleConnectedStreams(env)
+				.transform(transform)
+				.addSink(new DiscardingSink<>());
+	}
+
+	public static class MultiplyByTwoOperatorFactory extends AbstractStreamOperatorFactory<Long> {
+		@Override
+		public <T extends StreamOperator<Long>> T createStreamOperator(StreamOperatorParameters<Long> parameters) {
+			return (T) new MultiplyByTwoOperator(parameters);
+		}
+
+		@Override
+		public Class<? extends StreamOperator> getStreamOperatorClass(ClassLoader classLoader) {
+			return MultiplyByTwoOperator.class;
+		}
+	}
+
+	public static class MultiplyByTwoOperator extends AbstractStreamOperatorV2<Long> implements MultipleInputStreamOperator<Long> {
+		public MultiplyByTwoOperator(StreamOperatorParameters<Long> parameters) {
+			super(parameters, 2);
+		}
+
+		@Override
+		public List<Input> getInputs() {
+			return Arrays.asList(
+					new MultiplyByTwoInput(this, 1),
+					new MultiplyByTwoInput(this, 2));
+		}
+
+		private class MultiplyByTwoInput extends AbstractInput<Long, Long> {
+			public MultiplyByTwoInput(
+					AbstractStreamOperatorV2<Long> owner,
+					int inputId) {
+				super(owner, inputId);
+			}
+
+			@Override
+			public void processElement(StreamRecord<Long> element) {
+				output.collect(element.replace(element.getValue() * 2));
```

Comment:
Should object reuse be disabled for that?

Reply:
Object reuse is enabled by default in the benchmarks (`FlinkEnvironmentContext`).